README.md
# meshlab-pickedpoints

[![version](https://img.shields.io/pypi/v/meshlab-pickedpoints?style=flat-square)][pypi]
[![python versions](https://img.shields.io/pypi/pyversions/meshlab-pickedpoints?style=flat-square)][pypi]
[![license](https://img.shields.io/pypi/l/meshlab-pickedpoints?style=flat-square)][pypi]
[![coverage](https://img.shields.io/badge/coverage-100%25-brightgreen?style=flat-square)][coverage]
[![build](https://img.shields.io/circleci/project/github/lace/meshlab-pickedpoints/master?style=flat-square)][build]
[![code style](https://img.shields.io/badge/code%20style-black-black?style=flat-square)][black]

Read and write MeshLab picked point (`.pp`) files.

[pypi]: https://pypi.org/project/meshlab-pickedpoints/
[coverage]: https://github.com/lace/meshlab-pickedpoints/blob/master/.coveragerc#L2
[build]: https://circleci.com/gh/lace/meshlab-pickedpoints/tree/master
[docs build]: https://meshlab-pickedpoints.readthedocs.io/en/latest/
[black]: https://black.readthedocs.io/en/stable/

## Contribute

- Issue Tracker: https://github.com/lace/meshlab-pickedpoints/issues
- Source Code: https://github.com/lace/meshlab-pickedpoints

Pull requests welcome!


## Support

If you are having issues, please let us know.


## Acknowledgements

This library was based on code in [Lace][], which was refactored from legacy
code at Body Labs by [Alex Weiss][], with portions by [Eric Rachlin][],
[Paul Melnikow][], [Victor Alvarez][], and others. Later it was extracted
from the Body Labs codebase and open-sourced by [Guillaume Marceau][]. In
2018 it was [forked by Paul Melnikow][fork] and published as
[metabolace][fork pypi]. Thanks to a repository and package transfer from
Body Labs, the fork has been merged back into the original.

[lace]: http://github.com/lace/lace
[paul melnikow]: https://github.com/paulmelnikow
[jake beard]: https://github.com/jbeard4
[alex weiss]: https://github.com/algrs
[eric rachlin]: https://github.com/eerac
[victor alvarez]: https://github.com/yangmillstheory
[guillaume marceau]: https://github.com/gmarceau
[fork]: https://github.com/metabolize/lace
[fork pypi]: https://pypi.org/project/metabolace/

## License

The project is licensed under the two-clause BSD license.


CHANGELOG.md
# Changelog

# 4.1.0

- Support Python 3.7


# 4.0.0

BREAKING CHANGES:

- Remove `__version__` property.
- Require Python 3.8.1 or later.


# 3.0.0

BREAKING CHANGES:

- Adopt Metabolize/Curvewise named points JSON schema.
- Remove NumPy dependency; return lists instead.

Old:

```py
{
  "Femoral_epicon_med_lft": np.array([0.044259, 0.467733, -0.060032]),
  "Clavicale_lft": np.array([0.017893, 1.335375, 0.018390]),
  "Substernale": np.array([0.000625, 1.124424, 0.080930]),
}
```

New:
```py
[
  {"name": "Femoral_epicon_med_lft", "point": [0.044259, 0.467733, -0.060032]},
  {"name": "Clavicale_lft", "point": [0.017893, 1.335375, 0.018390]},
  {"name": "Substernale", "point": [0.000625, 1.124424, 0.080930]},
]
```

New features:

- Add picked points to JSON conversion script: `python -m meshlab_pickedpoints.cli`.


# 2.0.0

Upgrade vg dependency.


# 1.0.0

Version bump. No changes.


# 1.0.0b0

Initial release.


meshlab_pickedpoints/test_cli.py
import json
from click.testing import CliRunner
import pytest
from .cli import convert_landmarks
from .testing_fixtures import EXAMPLE_PICKED_POINTS_XML_STRING, EXAMPLE_POINTS
from .testing_helpers import assert_equal_points


@pytest.fixture
def write_tmp_picked_points(tmp_path):
    def _write_tmp_picked_points(xml_contents, basename="example.pp"):
        out_path = str(tmp_path / basename)
        with open(out_path, "w") as f:
            f.write(xml_contents)
        return out_path

    return _write_tmp_picked_points


def test_convert_landmarks_specified_output(tmp_path, write_tmp_picked_points):
    picked_points_path = write_tmp_picked_points(EXAMPLE_PICKED_POINTS_XML_STRING)
    output_path = tmp_path / "my_points.json"

    runner = CliRunner()
    result = runner.invoke(convert_landmarks, [picked_points_path, "-o", output_path])

    assert result.exit_code == 0
    assert result.output == ""

    with open(output_path, "r") as f:
        loaded = json.load(f)

    assert_equal_points(loaded, EXAMPLE_POINTS)


def test_convert_landmarks_default_output(tmpdir, write_tmp_picked_points):
    with tmpdir.as_cwd():
        picked_points_path = write_tmp_picked_points(EXAMPLE_PICKED_POINTS_XML_STRING)

        runner = CliRunner()
        result = runner.invoke(convert_landmarks, [picked_points_path])

        assert result.exit_code == 0
        assert result.output == ""

        with open("example.json", "r") as f:
            loaded = json.load(f)

        assert_equal_points(loaded, EXAMPLE_POINTS)


meshlab_pickedpoints/test_core.py
# Adapted from
# https://github.com/lace/lace/blob/d3c191dffaeedc14aafa4af031d74743de9e632d/lace/serialization/meshlab_pickedpoints.py
# https://github.com/lace/lace/blob/d3c191dffaeedc14aafa4af031d74743de9e632d/lace/test_meshlab_pickedpoints.py

from lxml import etree, objectify
import pytest
from ._core import dump, load, loads

example_xml_string = """
<!DOCTYPE PickedPoints>
<PickedPoints>
 <DocumentData>
  <DateTime time="16:00:00" date="2014-12-31"/>
  <User name="bodylabs"/>
  <DataFileName name="ignored.obj"/>
 </DocumentData>
 <point x="0.044259" y="0.467733" z="-0.060032" name="Femoral_epicon_med_lft"/>
<point x="0.017893" y="1.335375" z="0.01839" name="Clavicale_lft"/>
<point x="0.000625" y="1.124424" z="0.08093" name="Substernale"/>
</PickedPoints>
"""

example_points = [
    {
        "name": "Femoral_epicon_med_lft",
        "point": [0.044259, 0.467733, -0.060032],
    },
    {"name": "Clavicale_lft", "point": [0.017893, 1.335375, 0.018390]},
    {"name": "Substernale", "point": [0.000625, 1.124424, 0.080930]},
]


def assert_equal_points(first, second):
    assert len(first) == len(second)
    for first_item, second_item in zip(first, second):
        assert first_item["name"] == second_item["name"]
        assert first_item["point"] == second_item["point"]


def test_load():
    from io import StringIO

    sample_f = StringIO(example_xml_string)

    try:
        result = load(sample_f)
    finally:
        sample_f.close()

    assert_equal_points(result, example_points)


def test_loads():
    result = loads(example_xml_string)

    assert_equal_points(result, example_points)


def test_loads_error():
    xml_with_empty_point = """
    <!DOCTYPE PickedPoints>
    <PickedPoints>
    <DocumentData>
    <DateTime time="16:00:00" date="2014-12-31"/>
    <User name="bodylabs"/>
    <DataFileName name="ignored.obj"/>
    </DocumentData>
    <point x="0.044259" y="0.467733" z="-0.060032" name="Femoral_epicon_med_lft"/>
    <point x=" " y=" " z=" " name="Clavicale_lft"/>
    <point x="0.000625" y="1.124424" z="0.08093" name="Substernale"/>
    </PickedPoints>
    """
    expected_points = [
        {
            "name": "Femoral_epicon_med_lft",
            "point": [0.044259, 0.467733, -0.060032],
        },
        {"name": "Substernale", "point": [0.000625, 1.124424, 0.080930]},
    ]

    result = loads(xml_with_empty_point)

    assert_equal_points(result, expected_points)


def assert_equal_xml(value, expected):
    # This is fairly naive -- doesn't consider whitespace or that the points
    # could be in different order, and is strict about irrelevant parts like
    # the DateTime. But seems to work okay.
    value_normalized = etree.tostring(objectify.fromstring(value))
    expected_normalized = etree.tostring(objectify.fromstring(expected))

    assert value_normalized == expected_normalized


def test_dump():
    from io import StringIO

    result_f = StringIO()

    try:
        dump(example_points, result_f)
        result_str = result_f.getvalue()
    finally:
        result_f.close()

    assert_equal_xml(result_str, example_xml_string)


def test_dump_errors():
    from io import StringIO

    result_f = StringIO()

    try:
        with pytest.raises(ValueError, match="obj should be a list of points"):
            dump(-1, result_f)
    finally:
        result_f.close()

    try:
        with pytest.raises(
            ValueError,
            match="Expected keys to include point and optional name; got name, point, badkey",
        ):
            dump([{"name": "foo", "point": None, "badkey": -1}], result_f)
    finally:
        result_f.close()


dev.py
#!/usr/bin/env -S poetry run python

import os
import click
from executor import execute


def python_source_files():
    import glob

    include_paths = (
        glob.glob("*.py")
        + glob.glob("meshlab_pickedpoints/*.py")
        + glob.glob("meshlab_pickedpoints/**/*.py")
        + ["doc/"]
    )
    exclude_paths = []
    return [x for x in include_paths if x not in exclude_paths]


@click.group()
def cli():
    pass


@cli.command()
def install():
    execute("poetry install --sync --extras cli")


@cli.command()
def test():
    execute("pytest")


@cli.command()
def coverage():
    execute("pytest --cov=meshlab_pickedpoints")


@cli.command()
def coverage_report():
    execute("coverage html --fail-under=0")
    execute("open htmlcov/index.html")


@cli.command()
def lint():
    execute("flake8", *python_source_files())


@cli.command()
def black():
    execute("black", *python_source_files())


@cli.command()
def black_check():
    execute("black", "--check", *python_source_files())


@cli.command()
def doc():
    execute("rm -rf build/ doc/build/ doc/api/")
    execute("sphinx-build -W -b singlehtml doc doc/build")


@cli.command()
def doc_open():
    execute("open doc/build/index.html")


@cli.command()
def clean():
    execute("find . -name '*.pyc' -delete")
    execute("find . -name '__pycache__' -delete")


@cli.command()
def publish():
    execute("rm -rf dist/ build/")
    execute("poetry build")
    execute("twine upload dist/*")


if __name__ == "__main__":
    os.chdir(os.path.abspath(os.path.dirname(__file__)))
    cli()


meshlab_pickedpoints/_core.py
"""
Read and write a MeshLab picked points file. Files contain named points in 3D
space. A set of picked points can be saved alongside the mesh and loaded in
MeshLab (on the **Edit** menu, click **PickPoints**).
"""

# Adapted from
# https://github.com/lace/lace/blob/d3c191dffaeedc14aafa4af031d74743de9e632d/lace/serialization/meshlab_pickedpoints.py
# https://github.com/lace/lace/blob/d3c191dffaeedc14aafa4af031d74743de9e632d/lace/test_meshlab_pickedpoints.py

from io import StringIO
from xml.etree import ElementTree

__all__ = ["load", "dumps", "dump"]


def load(fp):
    """
    Deserialize the `.pp` file in `fp` to an object.

    Args:
        fp: An open file pointer.

    Return:
        dict: A dict with string keys and 3D NumPy points.
    """
    tree = ElementTree.parse(fp)

    points = []
    for e in tree.iter("point"):
        try:
            point = [float(e.attrib["x"]), float(e.attrib["y"]), float(e.attrib["z"])]
        except ValueError:
            # This may happen if landmarks are just spaces.
            continue
        points.append({"name": e.attrib["name"], "point": point})

    return points


def loads(xml_string):
    """
    Deserialize a `.pp` XML string to an object.

    Args:
        xml_string (str): The contents of a `.pp` file.

    Return:
        dict: A dict with string keys and 3D NumPy points.
    """
    f = StringIO(xml_string)
    try:
        return load(f)
    finally:
        f.close()


def dumps(points):
    """
    Render a list of points to a string containing the contents of a `.pp` file.

    Args:
        points (list): A list of dicts of the form `{"name": name, "point": point}`

    Return:
        str: A string containing the `.pp` file contents.
    """
    # TODO Maybe reconstruct this using xml.etree.
    if not isinstance(points, list):
        raise ValueError("obj should be a list of points")
    for point in points:
        if not set(point.keys()).issubset(set(["name", "point"])):
            raise ValueError(
                f"Expected keys to include point and optional name; got {', '.join(point.keys())}"
            )
        assert isinstance(point["point"], list)
        assert len(point["point"]) == 3

    points_xml_string = "\n".join(
        [
            '<point x="{}" y="{}" z="{}" name="{}"/>'.format(
                *point["point"], point["name"]
            )
            for point in points
        ]
    )

    return """
    <!DOCTYPE PickedPoints>
    <PickedPoints>
     <DocumentData>
      <DateTime time="16:00:00" date="2014-12-31"/>
      <User name="bodylabs"/>
      <DataFileName name="ignored.obj"/>
     </DocumentData>
     {}
    </PickedPoints>
    """.format(
        points_xml_string
    )


def dump(points, fp):
    """
    Render a list of points to an open file.

    Args:
        points (list): A list of dicts of the form `{"name": name, "point": point}`
        fp: An open file pointer.
    """
    xml_string = dumps(points)
    fp.write(xml_string)
